File: js/util/comparison.js
// Aggregiert äquivalente Ausdrücke quer über alle Dialekte.
// Strategie: gruppiere nach normalisierter Hochdeutsch-Übersetzung
// (mit Token-Overlap als Fallback für leichte Varianten).

import { ALLE_AUSDRUECKE } from '../../data/dialekte.js';
import { normalize } from './text.js';

function tokenSet(s) {
  return new Set(normalize(s).split(/\s+/).filter((w) => w.length >= 3));
}

// Jaccard-artige Ähnlichkeit über bereits berechnete Token-Sets.
function overlap(A, B) {
  if (!A.size || !B.size) return 0;
  // über das kleinere Set iterieren
  const [small, big] = A.size <= B.size ? [A, B] : [B, A];
  let hits = 0;
  for (const t of small) if (big.has(t)) hits++;
  return hits / Math.max(A.size, B.size);
}

// Gruppiert Ausdrücke nach ähnlicher Hochdeutsch-Übersetzung.
// minSize: nur Gruppen mit mindestens N Dialekten zurückgeben (default 2).
//
// Performance: statt jeden Kopf linear gegen ALLE Gruppen zu vergleichen
// (O(n²) + ständige Token-Set-Neuberechnung), nutzen wir
//   1. eine Exact-Map (normalisierter Kopf → Gruppe) für identische Köpfe,
//   2. einen invertierten Token-Index (Token → Gruppen), sodass nur Gruppen
//      verglichen werden, die mindestens ein Token teilen. Da ein Overlap
//      ≥ 0.6 zwingend ≥ 1 gemeinsames Token bedeutet, ist das verlustfrei.
//   3. pro Gruppe einmalig gecachte Token-Sets (headTokens).
export function buildComparison({ minSize = 2 } = {}) {
  const groups = [];
  const exact = new Map();      // normalisierter Kopf → Gruppe
  const tokenIndex = new Map(); // Token → Set<Gruppe>

  for (const a of ALLE_AUSDRUECKE) {
    const head = a.hochdeutsch.split(/[\/,–\-]+/)[0].trim();
    if (!head) continue;
    const norm = normalize(head);

    let target = exact.get(norm) || null;

    if (!target) {
      const tokens = tokenSet(head);
      // Kandidaten: nur Gruppen, die mindestens ein Token teilen.
      const candidates = new Set();
      for (const t of tokens) {
        const gs = tokenIndex.get(t);
        if (gs) for (const g of gs) candidates.add(g);
      }
      for (const g of candidates) {
        if (overlap(tokens, g.headTokens) >= 0.6) { target = g; break; }
      }
      if (!target) {
        target = { headRaw: head, head, kategorie: a.kategorie, items: [], headTokens: tokens };
        groups.push(target);
        exact.set(norm, target);
        for (const t of tokens) {
          let gs = tokenIndex.get(t);
          if (!gs) { gs = new Set(); tokenIndex.set(t, gs); }
          gs.add(target);
        }
      }
    }
    // Nur ein Ausdruck pro Dialekt pro Gruppe
    if (!target.items.some((x) => x.dialektId === a.dialektId)) {
      target.items.push(a);
    }
  }

  // Filtere + sortiere: mehr Dialekte = wichtiger
  return groups
    .filter((g) => g.items.length >= minSize)
    .sort((a, b) => b.items.length - a.items.length);
}

// Filter nach Kategorie.
export function filterByKategorie(groups, kategorie) {
  if (!kategorie || kategorie === 'all') return groups;
  return groups.filter((g) => g.kategorie === kategorie);
}

File: mobile/lib/util/comparison.dart
import '../data/models.dart';

/// Aggregiert äquivalente Ausdrücke quer über alle Dialekte — Port von
/// js/util/comparison.js. Gruppiert nach normalisierter Hochdeutsch-Übersetzung
/// (Token-Overlap ≥ 0.6 als Fallback für leichte Varianten).
///
/// Zur Performance bei vielen Tausend Ausdrücken nutzt diese Portierung einen
/// invertierten Token-Index: Statt jeden Kopf gegen ALLE bestehenden Gruppen zu
/// prüfen (O(n²)), werden nur Kandidaten verglichen, die mindestens ein Token
/// teilen — bei identischer „first-match"-Semantik wie das Web-Original.

typedef DialektAusdruck = ({Dialekt dialekt, Ausdruck ausdruck});

String _normalize(String s) {
  var t = s.toLowerCase();
  const map = {
    'ä': 'a', 'ö': 'o', 'ü': 'u', 'ß': 'ss',
    'á': 'a', 'à': 'a', 'â': 'a', 'é': 'e', 'è': 'e', 'ê': 'e',
  };
  map.forEach((k, v) => t = t.replaceAll(k, v));
  return t.trim();
}

Set<String> _tokenSet(String s) => _normalize(s)
    .split(RegExp(r'\s+'))
    .where((w) => w.length >= 3)
    .toSet();

double _similarity(String a, String b) {
  if (a == b) return 1;
  final setA = _tokenSet(a);
  final setB = _tokenSet(b);
  if (setA.isEmpty || setB.isEmpty) return 0;
  var hits = 0;
  for (final t in setA) {
    if (setB.contains(t)) hits++;
  }
  return hits / (setA.length > setB.length ? setA.length : setB.length);
}

/// Eine Gruppe gleichbedeutender Ausdrücke verschiedener Dialekte.
class CompareGroup {
  CompareGroup({required this.head, required this.kategorie});

  final String head; // Hochdeutsch-Kopf (Anzeige)
  final String kategorie;
  final List<DialektAusdruck> items = [];

  int get size => items.length;
}

/// Splittet auf den ersten Bedeutungs-Kopf (vor /, Komma, Gedanken-/Bindestrich).
final RegExp _headSplit = RegExp('[/,–-]+');

/// Gruppiert [pairs] nach ähnlicher Hochdeutsch-Übersetzung. Nur Gruppen mit
/// mindestens [minSize] Dialekten, absteigend nach Größe.
List<CompareGroup> buildComparison(
  List<DialektAusdruck> pairs, {
  int minSize = 2,
}) {
  final groups = <CompareGroup>[];
  final byExactHead = <String, int>{};
  final tokenIndex = <String, List<int>>{};

  for (final p in pairs) {
    final head = p.ausdruck.hochdeutsch.split(_headSplit).first.trim();
    if (head.isEmpty) continue;
    final tokens = _tokenSet(head);

    // Kandidaten in Erstell-Reihenfolge (exakter Kopf + Token-Index).
    final candidates = <int>{};
    final exact = byExactHead[head];
    if (exact != null) candidates.add(exact);
    for (final t in tokens) {
      final idxs = tokenIndex[t];
      if (idxs != null) candidates.addAll(idxs);
    }
    final sorted = candidates.toList()..sort();

    int? targetIdx;
    for (final gi in sorted) {
      if (_similarity(head, groups[gi].head) >= 0.6) {
        targetIdx = gi;
        break;
      }
    }

    if (targetIdx == null) {
      final gi = groups.length;
      final g = CompareGroup(head: head, kategorie: p.ausdruck.kategorie);
      g.items.add(p);
      groups.add(g);
      byExactHead.putIfAbsent(head, () => gi);
      for (final t in tokens) {
        (tokenIndex[t] ??= []).add(gi);
      }
    } else {
      final g = groups[targetIdx];
      // Nur ein Ausdruck pro Dialekt pro Gruppe.
      if (!g.items.any((x) => x.dialekt.id == p.dialekt.id)) {
        g.items.add(p);
      }
    }
  }

  final out = groups.where((g) => g.size >= minSize).toList();
  out.sort((a, b) => b.size - a.size);
  return out;
}

/// Filtert Gruppen nach Kategorie ('all'/leer → alle).
List<CompareGroup> filterByKategorie(
  List<CompareGroup> groups,
  String kategorie,
) {
  if (kategorie.isEmpty || kategorie == 'all') return groups;
  return groups.where((g) => g.kategorie == kategorie).toList();
}
